### PR TITLE
dist/common/scripts/scylla_util.py: GCP nvme: fail silently if you ca…

### LINKED ISSUE #8032: scylla_io_setup on GKE: found more than one disk mounted at root
*Installation details*
Scylla version: 4.3.0
Cluster size: 3
OS: GKE (over GCE) + scylla-operator using Helm

Platform (physical/VM/cloud instance type/docker): GKE cloud instance with K8s v1.16.15-gke.6000
Hardware: n1-standard-2 (vcpu=2 memory=7.5G)
Disks: SSD persistent disk (pd-ssd), no local SSD

Logs when starting the first Scylla instance:
```
│ {"L":"INFO","T":"2021-02-03T17:20:23.602Z","N":"member_controller","M":"Starting reconciliation...","_trace_id":"hhrl_QEtQcmJlorMijTAxQ"}
│ running: (['/opt/scylladb/scripts/scylla_dev_mode_setup', '--developer-mode', '0'],)                                                                                                                                                             │
│ running: (['/opt/scylladb/scripts/scylla_cpuset_setup', '--cpuset', '0-1'],)                                                                                                                                                                     │
│ running: (['/opt/scylladb/scripts/scylla_io_setup'],)                                                                                                                                                                                            │
│ Problem when parsing disks from OS:                                                                                                                                                                                                              │
│ found more than one disk mounted at root                                                                                                                                                                                                         │
│ Traceback (most recent call last):                                                                                                                                                                                                               │
│   File "/opt/scylladb/scripts/libexec/scylla_io_setup", line 239, in <module>                                                                                                                                                                    │
│     if idata.is_recommended_instance():                                                                                                                                                                                                          │
│   File "/opt/scylladb/scripts/scylla_util.py", line 311, in is_recommended_instance                                                                                                                                                              │
│     diskSize = self.firstNvmeSize                                                                                                                                                                                                                │
│   File "/opt/scylladb/scripts/scylla_util.py", line 290, in firstNvmeSize                                                                                                                                                                        │
│     ephemeral_disks = self.getEphemeralOsDisks()                                                                                                                                                                                                 │
│   File "/opt/scylladb/scripts/scylla_util.py", line 177, in getEphemeralOsDisks                                                                                                                                                                  │
│     return self.os_disks[self.EPHEMERAL]                                                                                                                                                                                                         │
│   File "/opt/scylladb/scripts/scylla_util.py", line 169, in os_disks                                                                                                                                                                             │
│     nvmes_present = self._non_root_nvmes()                                                                                                                                                                                                       │
│   File "/opt/scylladb/scripts/scylla_util.py", line 155, in _non_root_nvmes                                                                                                                                                                      │
│     raise Exception("found more than one disk mounted at root ".format(root_dev_candidates))                                                                                                                                                     │
│ Exception: found more than one disk mounted at root                                                                                                                                                                                              │
│ failed!                                                                                                                                                                                                                                          │
│ Traceback (most recent call last):                                                                                                                                                                                                               │
│   File "/docker-entrypoint.py", line 27, in <module>                                                                                                                                                                                             │
│     setup.io()                                                                                                                                                                                                                                   │
│   File "/scyllasetup.py", line 67, in io                                                                                                                                                                                                         │
│     self._run(['/opt/scylladb/scripts/scylla_io_setup'])                                                                                                                                                                                         │
│   File "/scyllasetup.py", line 37, in _run                                                                                                                                                                                                       │
│     subprocess.check_call(*args, **kwargs)                                                                                                                                                                                                       │
│   File "/opt/scylladb/python3/lib64/python3.8/subprocess.py", line 364, in check_call                                                                                                                                                            │
│     raise CalledProcessError(retcode, cmd)                                                                                                                                                                                                       │
│ subprocess.CalledProcessError: Command '['/opt/scylladb/scripts/scylla_io_setup']' returned non-zero exit status 1.
```

Result of the `mount` command on the instance:
```
overlay on / type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay2/l/32O6Y3GND5YVHIDEYAYQALKJYW:/var/lib/docker/overlay2/l/DZOPRJ3YPUGL2LLOWS7HTYEPJJ:/var/lib/docker/overlay2/l/QVJYLMFUGLTNU5CFRYOIQ4H6CH:/var/lib/docker/overlay2/l/Y57AOZDPCTSZS6IJOMGJLEUODO:/var/lib/docker/overlay2/l/IOFSLW65XQUNHBD4EBVOQMQTH5:/var/lib/docker/overlay2/l/C3XIQX6HVNY5S7BWYRL5SLVJFJ:/var/lib/docker/overlay2/l/E4UXSM63JPMLKALZIZIDHZOPVW:/var/lib/docker/overlay2/l/IIN6QTH56OTAT7I2A3EOPIKKQ6:/var/lib/docker/overlay2/l/UVEO2RDQGFNMMWYADP3NFDSZMH:/var/lib/docker/overlay2/l/DWMKLR6XF7CNECA6MXMA7UJSZC:/var/lib/docker/overlay2/l/NMY32XTDDO5R5N22TA264CADUV:/var/lib/docker/overlay2/l/WABZOCGMAGPU766T7R26GYXVZD:/var/lib/docker/overlay2/l/O2FE76MNRKFNTB62MXESLTWDMU:/var/lib/docker/overlay2/l/GHPXZXQAXE32NDYDPXOQWFIEVR:/var/lib/docker/overlay2/l/U5P2HJWRPUDQL7ROXHEBTXQSCX:/var/lib/docker/overlay2/l/BIR2EN3J2LHADN3YMZKWGRXT77:/var/lib/docker/overlay2/l/XIE7AYJOVLPZJTAD5V4OQIGYFX:/var/lib/docker/overlay2/l/LK2AII4JXD6CSX2GSUKZTVWDCE:/var/lib/docker/overlay2/l/JXFOQHNX6HMTD25LWOPWFDAL5X:/var/lib/docker/overlay2/l/C3AOL3MSRJFGLPIMXTO7VUZLOE:/var/lib/docker/overlay2/l/PCVFAEQXIIFYYMQ4O4JZVRO52R:/var/lib/docker/overlay2/l/2KEODVSTBHQQUM6PO7CCD4XCVE,upperdir=/var/lib/docker/overlay2/a14b6435f3fd8e0d205f2fa6b5de13210d2f13d827c0a0a855e15c5dbeb2e572/diff,workdir=/var/lib/docker/overlay2/a14b6435f3fd8e0d205f2fa6b5de13210d2f13d827c0a0a855e15c5dbeb2e572/work)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev type tmpfs (rw,nosuid,size=65536k,mode=755)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=666)
sysfs on /sys type sysfs (ro,nosuid,nodev,noexec,relatime)
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,relatime,mode=755)
cgroup on /sys/fs/cgroup/systemd type cgroup (ro,nosuid,nodev,noexec,relatime,xattr,name=systemd)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (ro,nosuid,nodev,noexec,relatime,hugetlb)
cgroup on /sys/fs/cgroup/perf_event type cgroup (ro,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/blkio type cgroup (ro,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/devices type cgroup (ro,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/cpuset type cgroup (ro,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (ro,nosuid,nodev,noexec,relatime,cpu,cpuacct)
cgroup on /sys/fs/cgroup/freezer type cgroup (ro,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/pids type cgroup (ro,nosuid,nodev,noexec,relatime,pids)
cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (ro,nosuid,nodev,noexec,relatime,net_cls,net_prio)
cgroup on /sys/fs/cgroup/memory type cgroup (ro,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/rdma type cgroup (ro,nosuid,nodev,noexec,relatime,rdma)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
/dev/sda1 on /mnt/shared type ext4 (ro,relatime,commit=30)
/dev/sda1 on /mnt/scylla-config type ext4 (ro,relatime,commit=30)
tmpfs on /mnt/scylla-client-config type tmpfs (ro,relatime)
/dev/sda1 on /dev/termination-log type ext4 (rw,relatime,commit=30)
/dev/sda1 on /etc/resolv.conf type ext4 (rw,nosuid,nodev,relatime,commit=30)
/dev/sda1 on /etc/hostname type ext4 (rw,nosuid,nodev,relatime,commit=30)
/dev/sda1 on /etc/hosts type ext4 (rw,relatime,commit=30)
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
/dev/sdb on /var/lib/scylla type ext4 (rw,relatime)
tmpfs on /run/secrets/kubernetes.io/serviceaccount type tmpfs (ro,relatime)
proc on /proc/bus type proc (ro,relatime)
proc on /proc/fs type proc (ro,relatime)
proc on /proc/irq type proc (ro,relatime)
proc on /proc/sys type proc (ro,relatime)
proc on /proc/sysrq-trigger type proc (ro,relatime)
tmpfs on /proc/acpi type tmpfs (ro,relatime)
tmpfs on /proc/kcore type tmpfs (rw,nosuid,size=65536k,mode=755)
tmpfs on /proc/keys type tmpfs (rw,nosuid,size=65536k,mode=755)
tmpfs on /proc/timer_list type tmpfs (rw,nosuid,size=65536k,mode=755)
tmpfs on /proc/scsi type tmpfs (ro,relatime)
tmpfs on /sys/firmware type tmpfs (ro,relatime)
```

The line responsible for validating the number of disks is `if len(root_dev_candidates) != 1:`, the error is a bit misleading because in my case I suppose the number of disks found is actually `0` and not `> 1`.

Is running without local SSD unsupported on GKE?

### PR DESCRIPTION
**Pull request by tarzanek**
…nnot figure

out nvme count, or cannot figure out their size and fallback to iotune

fixes #8032

### CODE REVIEW

**Comment by tarzanek** [comment]
tested on @zimnx GKE setup (so no nvme exposed) + n2-standard-2 + 2 nvmes

File: dist/common/scripts/scylla_util.py [hunk 1]
```diff
@@ -154,20 +154,27 @@ def __instance_metadata(self, path, recursive=False):
...
 
-        root_dev = root_dev_candidates[0].device
+        root_devs = [x.device for x in root_dev_candidates]
         # if root_dev.startswith("/dev/mapper"):
         #     raise Exception("mapper used for root, not checking if nvme is used ".format(root_dev))
 
-        nvmes_present = list(filter(nvme_re.match, os.listdir("/dev")))
-        return {self.ROOT: [root_dev], self.EPHEMERAL: [x for x in nvmes_present if not root_dev.startswith(os.path.join("/dev/", x))]}
+        nvmes_present = list(filter(nvme_re.match, os.listdir("/dev")))   # it can be empty too - e.g. in k8s
+        return {self.ROOT: [root_devs], self.EPHEMERAL: [x for x in nvmes_present if not self.is_in_root_devs(x, root_devs)]}
```
Comment (zimnx) [inline dist/common/scripts/scylla_util.py:177]: `self.ROOT` type is `[][]string` (array of array of strings), where it was an `[]string` previously. Wouldn't it cause any issues?

File: dist/common/scripts/scylla_util.py [hunk 2]
```diff
@@ -154,20 +154,27 @@ def __instance_metadata(self, path, recursive=False):
...
     def _non_root_nvmes(self):
         """get list of nvme disks from os, filter away if one of them is root"""
         nvme_re = re.compile(r"nvme\d+n\d+$")
 
         root_dev_candidates = [x for x in psutil.disk_partitions() if x.mountpoint == "/"]
-        if len(root_dev_candidates) != 1:
-            raise Exception("found more than one disk mounted at root ".format(root_dev_candidates))
+        #it can be empty too - e.g. in k8s
+        #if len(root_dev_candidates) != 1:
+        #    raise Exception("found more than one disk mounted at root ".format(root_dev_candidates))
```
Comment (avikivity) [inline dist/common/scripts/scylla_util.py:171]: Why are you adding comments?

File: dist/common/scripts/scylla_util.py (hunk 1 above)
Comment (avikivity) [inline dist/common/scripts/scylla_util.py:177]: Is this related to the patch subject?

File: dist/common/scripts/scylla_util.py [hunk 3]
```diff
@@ -296,13 +303,17 @@ def firstNvmeSize(self):
         """return the size of first non root NVME disk in GB"""
         if self.__firstNvmeSize is None:
             ephemeral_disks = self.getEphemeralOsDisks()
-            firstDisk = ephemeral_disks[0]
-            firstDiskSize = self.get_file_size_by_seek(os.path.join("/dev/", firstDisk))
-            firstDiskSizeGB = firstDiskSize/1024/1024/1024
-            if firstDiskSizeGB >= self.GCP_NVME_DISK_SIZE_2020:
-                self.__firstNvmeSize = firstDiskSizeGB
+            if len(ephemeral_disks)>0:
```
Comment (avikivity) [inline dist/common/scripts/scylla_util.py:306]: Please use normal Python coding style.

**Comment by avikivity** [comment]
Try to make your patches less hostile to reviewers.

File: dist/common/scripts/scylla_util.py (hunk 2 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:171]: will remove all comments

File: dist/common/scripts/scylla_util.py (hunk 1 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:177]: nope, it wasn't used anywhere, so the change of return type is safe

Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:177]: definitely, the logic changed, now for every root device we need to check if it belongs to an nvme and filter that nvme out from available ones

reason is simple - overlay mounts in K8S create multiple / mounts

File: dist/common/scripts/scylla_util.py (hunk 3 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:306]: done

**Comment by tarzanek** [comment]
thanks for review @avikivity @zimnx 
@avikivity I have a conflict between writing+testing and then adjusting the code for reviewers (where the biggest pain point is the need to retest the patch modification, despite the fact I have most of pieces for testing automated), but I will try to improve, I understand it's a pain to review (and here I definitely forgot to do auto format, sorry for that)

**Review by zimnx — APPROVED** [review]
LGTM

File: dist/common/scripts/scylla_util.py (hunk 1 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:172]: actually just double thinking about it it doesn't make sense to wrap it again in []
will fix

Comment (tarzanek) [reply dist/common/scripts/scylla_util.py]: fixed

**Comment by zimnx** [comment]
@tarzanek @avikivity @bentsi friendly reminder, we need it Scylla Operator to support GCP.

**Comment by avikivity** [comment]
Why is it okay to fail silently? Usually it's a bug. Suppose we have a bug there, it will fail silently, and we won't notice.

**Comment by tarzanek** [comment]
It's OK since it will fall back to iotune (and if iotune fails, THEN we should get a bug filed).
Which is what we want in unknown configs/situations @avikivity  (unless you'll tell me we don't want it and want to have such bugs as this every time something little changes).
In a known situation/config/hw we will use the pre-set variables for known disk setup, otherwise use iotune, which won't break the scylla setup due to something we can recover from.
(software should be able to "self heal" and continue, shouldn't it?)

**Comment by avikivity** [comment]
> It's OK since it will fall back to iotune (and if iotune fails, THEN we should get a bug filed).
> Which is what we want in unknown configs/situations @avikivity (unless you'll tell me we don't want it and want to have such bugs as this every time something little changes).
> In a known situation/config/hw we will use the pre-set variables for known disk setup, otherwise use iotune, which won't break the scylla setup due to something we can recover from.
> (software should be able to "self heal" and continue, shouldn't it?)

It depends on the nature of the problem. If we detected an instance we don't know, we can proceed with iotune.

If the software failed due to some bug, at the very least it should fail loudly (so that we know about it). Whether we should continue with iotune is debatable. On the one hand, it allows the user to proceed. On the other hand, it can compound the failure. It really depends on why we failed in the first place.

**Comment by tarzanek** [comment]
would it make sense to then at least add logging @avikivity ?

**Comment by avikivity** [comment]
It may make sense, I can't tell, since I can't find out if you're trying to recover from a bug in the script, a transient error, or something else.

**Comment by tarzanek** [comment]
@avikivity transient error in this case
also the code later can handle the disk size 0 better than an exception
I will add logging where needed

**Comment by tarzanek** [comment]
well ... all the code paths eventually end
at
`logging.warning("First nvme is smaller than lowest expected size. ".format(firstDisk))`

so we will log the failure anyways and continue ...
at https://github.com/scylladb/scylla/pull/8444/files#diff-9fe4dbe99d9ee03a478c9b25175e06bbd4cef9303c0ab1b9bdeea6e52d284cd4R309
we could start throwing and blocking

(but then https://github.com/scylladb/scylla/pull/8444/files#diff-9fe4dbe99d9ee03a478c9b25175e06bbd4cef9303c0ab1b9bdeea6e52d284cd4L319 later will just fallback to iotune and to me it is OK, the numbers later can be investigated if a problem will appear in load, but the expectation is that iotune will want to produce "good" numbers (and latest 4.3/4.4 looks like it is producing numbers close to reality))

**Comment by avikivity** [comment]
Sorry, I just don't understand. Please explain the scenario, how the existing code fails, and how the new code is better.

**Comment by tarzanek** [comment]
old code didn't accept multiple root mount points,
new code does

old code failed for missing nvmes (so if you run it on PD-SSD it will fail),
new code continues in such case with nvme disk size 0, which makes later checks fallback to iotune

btw. this second problem makes this a backport candidate, otherwise recent enterprise will fail on non-nvme GCP instances

**Comment by tarzanek** [comment]
reason for fallback to iotune for Macieks case is due to k8s, so the measured IOPS will be smaller than on host directly

**Comment by tarzanek** [comment]
interesting, after I saw https://github.com/scylladb/scylla/issues/8055 , it seems there is more to this change, so making a generic class that the 3 cloud implementations will inherit seems like even better idea than before

**Comment by slivne** [comment]
This is blocking using Scylla on GKE (the last scylla version that works is Scylla 4.2)

Before we rewrite all the code we need a working solution and then do iterations to approve it consolidate all clouds etc.

@tarzanek / @avikivity

**Comment by tarzanek** [comment]
@slivne this is working, it affects only GCP and it was tested by me on normal VMs(with and without NVMEs) and in GKE in Maciejs setup

**Comment by tarzanek** [comment]
@slivne can the GKE guys just take this patch from some custom build and use it in their tests?

File: dist/common/scripts/scylla_util.py [hunk 4]
```diff
@@ -154,20 +154,22 @@ def __instance_metadata(self, path, recursive=False):
         return curl(self.META_DATA_BASE_URL + path + "?recursive=%s" % str(recursive).lower(),
                     headers={"Metadata-Flavor": "Google"})
 
+    def is_in_root_devs(self, x, root_devs):
+        for root_dev in root_devs:
+            if root_dev.startswith(os.path.join("/dev/", x)):
+                return True
+        return False
```
Comment (avikivity) [inline dist/common/scripts/scylla_util.py:161]: I don't understand what this is calculating.

Say x is `sda` and `root_devs` is `[/dev/sda1, /dev/sdb1]`. The logic will return True. But what does it mean?

Mixing plain names with paths is confusing, and using startswith() on file names is liable to get wrong results if two devices have names that partially match.

File: dist/common/scripts/scylla_util.py [hunk 5]
```diff
@@ -154,20 +154,22 @@ def __instance_metadata(self, path, recursive=False):
...
-        if len(root_dev_candidates) != 1:
-            raise Exception("found more than one disk mounted at root ".format(root_dev_candidates))
 
-        root_dev = root_dev_candidates[0].device
-        # if root_dev.startswith("/dev/mapper"):
-        #     raise Exception("mapper used for root, not checking if nvme is used ".format(root_dev))
+        root_devs = [x.device for x in root_dev_candidates]
 
         nvmes_present = list(filter(nvme_re.match, os.listdir("/dev")))
-        return {self.ROOT: [root_dev], self.EPHEMERAL: [x for x in nvmes_present if not root_dev.startswith(os.path.join("/dev/", x))]}
```
Comment (avikivity) [inline dist/common/scripts/scylla_util.py:170]: Oh, I see you just copied the existing bad logic.

**Comment by avikivity** [comment]
I can't make any sense of this. @penberg please review it.

File: dist/common/scripts/scylla_util.py (hunk 5 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:170]: yep, same logic is in AWS, where it originated

File: dist/common/scripts/scylla_util.py (hunk 4 above)
Comment (penberg) [reply dist/common/scripts/scylla_util.py:161]: @tarzanek Can we change `_non_root_nvmes` to use `glob()` instead of `os.listdir()` to get the full path as follows (the regex also needs tweaking a bit):

```python
nvmes_present = list(filter(nvme_re.match, glob(os.path.join("/dev", "*"))))
```

We can then simplify `is_in_root_devs` to be:

```
def is_in_root_devs(self, x, root_devs):
    return x in root_devs
```

File: dist/common/scripts/scylla_util.py [hunk 6]
```diff
@@ -296,13 +298,17 @@ def firstNvmeSize(self):
         """return the size of first non root NVME disk in GB"""
         if self.__firstNvmeSize is None:
             ephemeral_disks = self.getEphemeralOsDisks()
-            firstDisk = ephemeral_disks[0]
-            firstDiskSize = self.get_file_size_by_seek(os.path.join("/dev/", firstDisk))
-            firstDiskSizeGB = firstDiskSize/1024/1024/1024
-            if firstDiskSizeGB >= self.GCP_NVME_DISK_SIZE_2020:
-                self.__firstNvmeSize = firstDiskSizeGB
+            if len(ephemeral_disks) > 0:
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:301]: Is this change part of the same issue or a different bug?

File: dist/common/scripts/scylla_util.py (hunk 4 above)
Comment (penberg) [reply dist/common/scripts/scylla_util.py:161]: @avikivity At high-level, the `_non_root_nvmes` function is attempting to provide a list of NVMe device files that are *not* mounted to "/". I agree that the code is confusing and bit fragile.

File: dist/common/scripts/scylla_util.py (hunk 6 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:301]: same issue, if no nvmes found, we shall still continue and not fail

File: dist/common/scripts/scylla_util.py (hunk 4 above)
Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:161]: will fix and retest
